---
layout: page
title: Learning Objectives and Course Pre-requisites
---

# Learning Objectives

Course participants will go through a series of experiences that utimately lead to create enhanced capabilities to:

1. List broad characteristics of NGS technologies and choose adequate sequencing for your biological question
2. Have a broad overview of the steps in the analysis of RNA-Seq differential expression experiments
3. Assess the general quality of the raw data from the sequencing facility
4. Do simple processing operations in the raw data to improve its quality
5. Generate alignments against a reference genome
6. Assess the general quality of the alignments and detect possible problems
7. Generate tables of counts using the alignment and a reference gene annotation
8. Generate lists of differentially expressed genes, at least for a simple pairwise comparison
9. Understand specificies of differential gene expression in the case of single-cell RNAseq
10. Perform simple functional enrichment analysis and understand the concepts behind them

For this, we are providing small example datasets and exercises that participants can use to learn.

<br/>

# Course Pre-requisites

Familiarity with elementary statistics and a few basics of scripting in R will be helpful.

Please have a look at the following resources and gauge your ability to use R in statitics at the basic level: [Coursera videos](http://blog.revolutionanalytics.com/2012/12/coursera-videos.html); [Introduction to r](http://bitesizebio.com/webinar/20600/beginners-introduction-to-r-statistical-software)

Basic Unix command line skills, such as being able to navigate in a directory tree and copy files. See, for example, ["Session 1" of the Software Carpentry training for a Unix introduction](http://bioinformatics-core-shared-training.github.io/shell-novice/).

<br/>

### Back

Back to [first page](https://maccardoso.github.io/ADER18S/).
