# Introduction

In [Han, Xiaoping, et al. "Mapping the mouse cell atlas by Microwell-seq." Cell 172.5 (2018)](https://doi.org/10.1016/j.cell.2018.02.001) a scRNA-seq dataset of approximately 400,000 cells isolated from all major mouse organs was generated by Microwell-seq. Expression matrices for all samples are available [here](https://figshare.com/s/865e694ad06d5857db4b).

In the previous section of this course we processed the raw fastq files for a sample of cells from mouse lung in order to obtain an UMI count matrix for all mouse genes. Because we don't know how many cells are really in the sample, we constructed a matrix containing the most abundant 10,000 barcodes.

In this tutorial we will perform the steps necessary to go from the raw expression matrix to a list of clusters representing transcriptionally distinct cell sub-populations and a list of marker genes associated with each identified cluster using the web-application D-cellerate.

## D-cellerate

One of the most popular and widely used packages for scRNA-seq analysis is called *Seurat* (https://satijalab.org/seurat/). It is an *R* package that contains functions to perform filtering, quality control, normalization, dimensionality reduction, clustering and differential expression analysis of scRNA-seq. However, its use requires some knowledge of the *R* programming language, and thus can be difficult for non-experts.

To alieviate this problem, we are developing a web-application that exposes most *Seurat* functions in a user-friendly and interactive way.

To launch the application open the file _app.R_ in the D-cellerate directory using RStudio, and press "Run App" in the top right corner. Alternatively you can launch the application by running the following commands on the terminal.

```
cd ~/Desktop/singlecell_practical/D-cellerate
Rscript -e "shiny::runApp(launch.browser=TRUE)"
```

# Loading the raw UMI matrix

First we load an UMI matrix to analyse. In this tutorial we are going the lung sample previously described. For convenience it is available as an built-in dataset into the application.

In the **Import Data** section, select **Built-in dataset** and choose the ""*mca_lung_10k*"" dataset. On the right pane the *Summary* tab shows the number of columns (barcodes) and rows (genes) in the dataset.

**Question**: How many genes and barcodes are quantified in this raw UMI matrix?

<details><summary><b>Click Here to see the answer</b></summary> 16566 genes and 10000 barcodes. </details>

# Filtering the matrix

Move to the **Filter** section by selecting it in the left pane. On the right we find the **barcode plot** that shows the total number of UMI counts per barcode in the raw UMI matrix (as we have seen in the *Cellranger* analysis).

![](images/barcode-plot.png)

**Question**: What can you conclude from the above representation? How many of the top barcodes would you keep for further analysis?

<details><summary><b>Click Here to see the answer</b></summary>

There appears to be a drop in the total number of UMI counts after the first 1,000 barcodes. However, unlike what we saw in the 10x dataset, the separation between an empty GEM and a GEM containing a cell is less clear. This could be due to the presence of ambient RNA in the sample.


In the original study, taking into account the full set of 91 samples, the authors selected a threshold of 500 UMI counts to select barcodes for further analysis. Thus we are left with 2684 cells for further analysis.


</details>
